### PR TITLE
Feature/ts-rb-purchase-date-util #11

### PR DESCRIPTION
**Pull request by flbarfield**
## Description

In order to advise users about when to purchase things, the app needs to be able to calculate that guess and store it a future date.

The calculateEstimate function from the @the-collab-lab/shopping-list-utils module will help us know how many days in the future a purchase should happen. In order to use calculateEstimate, we need to know how many days have passed since the time the item was last purchased.

## Related Issue

Issue #11 

## Acceptance Criteria

- [ ]  When the user purchases an item, the item’s dateNextPurchased property is calculated using the calculateEstimate function and saved to the Firestore database
    - [ ]  dateNextPurchased is saved as a date, not a number

- [ ]  A getDaysBetweenDates function is exported from utils/dates.js and imported into api/firebase.js
    - [ ]  This function takes two JavaScript Dates and returns the number of days that have passed between them

## Type of Changes

Feature

### Before

N/A

### After

N/A

## Testing Steps / QA Criteria

1. When a list item is clicked, a new value in dateNextPurchased should be entered into the database. This could be checked by looking at the individual item that was clicked, and seeing if the value has changed upon that click. 
    -  Until this is fully implemented, the dateNextPurchase that the item should be updating to is currently being console logged

2. The other half of this issue was an implementation of the getDaysBetweenDates function that will reside within the dates.js/utils file. When given two firebase-date formatted items, it will return the number of days that have passed.

### CODE REVIEW

File: src/utils/dates.js [hunk 1]
```diff
@@ -10,3 +10,27 @@ const ONE_DAY_IN_MILLISECONDS = 86400000;
...
+	// error on new item creation.
+	if (!startDate) {
+		return;
+	}
+
+	// the .toDate method converts date formatting to milliseconds.
+	// divinding these days by ONE_DAY_IN_MILLISECONDS, gives you the total
+	// ammount of days after the difference is established.
+	if (startDate && endDate) {
+		const difference = startDate.toDate() - endDate.toDate();
```
Comment (bdanzer) [inline src/utils/dates.js]: I would improve the naming for increased readability. Each of these variables should be clear enough even for a non-coder to understand. What is the difference here?

File: src/utils/dates.js [hunk 2]
```diff
@@ -10,3 +10,27 @@ const ONE_DAY_IN_MILLISECONDS = 86400000;
...
+		const calculation = Math.floor(difference / ONE_DAY_IN_MILLISECONDS);
```
Comment (bdanzer) [inline src/utils/dates.js]: Same note on readability, what is calculation?

File: src/utils/dates.js [hunk 3]
```diff
@@ -10,3 +10,27 @@ const ONE_DAY_IN_MILLISECONDS = 86400000;
 export function getFutureDate(offset) {
 	return new Date(Date.now() + offset * ONE_DAY_IN_MILLISECONDS);
 }
+
+export function getDaysBetweenDates(startDate, endDate) {
+	// New items are created without a startDate. This avoids null type
+	// error on new item creation.
+	if (!startDate) {
+		return;
```
Comment (bdanzer) [inline src/utils/dates.js:20]: I would suggest still returning something here

File: src/components/ListItem.jsx [hunk 4]
```diff
@@ -27,10 +31,38 @@ export function ListItem({ item, listPath }) {
 		mutationFn: markAsPurchased,
 	});
 
+	const previousEstimate = getDaysBetweenDates(
+		dateNextPurchased,
+		dateLastPurchased,
+	);
+
+	function estimateAddedToNow() {
```
Comment (bdanzer) [inline src/components/ListItem.jsx]: I would export this function from `utils/dates.js` and turn this into a pure function

Comment (bdanzer) [reply src/components/ListItem.jsx]: Also from a readability standpoint I am having a hard time understanding what this function does.

File: src/utils/dates.js [hunk 5]
```diff
@@ -10,3 +10,27 @@ const ONE_DAY_IN_MILLISECONDS = 86400000;
...
+	// the .toDate method converts date formatting to milliseconds.
+	// divinding these days by ONE_DAY_IN_MILLISECONDS, gives you the total
+	// ammount of days after the difference is established.
+	if (startDate && endDate) {
+		const difference = startDate.toDate() - endDate.toDate();
+		const calculation = Math.floor(difference / ONE_DAY_IN_MILLISECONDS);
+
+		// if number cannot divide evenly into a number more than 1, than less than a day has
+		// passed. Defaults to 0.
+		if (calculation <= 0) {
```
Comment (bdanzer) [inline src/utils/dates.js]: alternative way to write:

```javascript
if (calculation <= 0) {
    return 0;
} 

return calculation;
```

File: src/components/ListItem.jsx [hunk 6]
```diff
@@ -27,10 +31,38 @@ export function ListItem({ item, listPath }) {
...
+
+		//adding milliseconds to date
+		currentDate.setTime(currentDate.getTime() + estDaysToMilli);
+
+		const updatedTimeSTamp = Timestamp.fromDate(currentDate);
+
+		// this should be the final conversion to firebase's date format.
+		console.log(updatedTimeSTamp.toDate());
+		return updatedTimeSTamp.toDate();
+	}
```
Comment (adidalal) [inline src/components/ListItem.jsx]: 💯 to moving out the date manipulation bit to utils.

File: src/utils/dates.js [hunk 7]
```diff
@@ -10,3 +10,27 @@ const ONE_DAY_IN_MILLISECONDS = 86400000;
 export function getFutureDate(offset) {
 	return new Date(Date.now() + offset * ONE_DAY_IN_MILLISECONDS);
 }
+
+export function getDaysBetweenDates(startDate, endDate) {
```
Comment (adidalal) [inline src/utils/dates.js]: consider keeping your functions atomic and doing "business logic" in firebase.js to compose them

Here's one thought on how to split up the functions: getFutureDate / getDaysBetweenDates / isMoreThanADayOld

File: src/api/firebase.js
Comment (adidalal) [inline src/api/firebase.js:1]: You would need to modify `updateItem` otherwise your calculations won't actually get persisted

**Review by adidalal — CHANGES_REQUESTED** [review]

File: src/api/firebase.js [hunk 8]
```diff
@@ -188,8 +191,41 @@ export async function updateItem(listPath, { itemId, totalPurchases }) {
...
+	/**
+	 * Calculate the days since the last purchase,
+	 * considering either date last purchased or the created date
+	 */
+	const daysSinceLastPurchase = getDaysBetweenDates(
+		today,
+		dateLastPurchased ? dateLastPurchased?.toDate() : today,
+	);
+
+	const smartNextPurchaseEstimate = calculateEstimate(
```
Comment (jtkabenni) [inline src/api/firebase.js]: this should be `let`

**Review by jtkabenni — COMMENTED** [review]
Hey @flbarfield and @TishShaw , getting this error when trying to mark an item as purchased, think something in your firebase function is a const when it should be let. 
Screenshot 2024-03-09 at 5 00 03 PM

File: vite.config.js [hunk 9]
```diff
@@ -60,6 +59,7 @@ export default defineConfig({
 	plugins: [react(), svgr({ exportAsDefault: true }), VitePWA(PWAConfig)],
 	server: { open: true, port: 3000 },
 	test: {
+		pool: 'vmThreads',
```
Comment (bdanzer) [inline vite.config.js:62]: NOTE: I added this to fix our tests 😄

File: src/api/firebase.js [hunk 10]
```diff
@@ -197,8 +200,41 @@ export async function updateItem(listPath, { itemId, totalPurchases }) {
 
 	const itemDoc = doc(db, listPath, 'items', itemId);
 
+	const today = new Date();
```
Comment (bdanzer) [inline src/api/firebase.js]: I created a PR to showcase this but I would try to separate logic into separate functions. It's a bit easier to show in a separate PR: 

https://github.com/the-collab-lab/tcl-67-smart-shopping-list/pull/47

**Pull request merged**